impl/common/data/record/mongo/distinct_cursor.go

```diff
@@ -0,0 +1,195 @@
+package mongorec
+
+import (
+	"context"
+	"fmt"
+	"github.com/samber/lo"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"kantoku/common/data/record"
+)
+
+type DistinctCursor struct {
+	skip    int
+	limit   int
+	filters [][]record.Entry
+	sorters []record.Sorter
+	masks   []record.Mask
+	keys    []string
+	storage *Storage
+}
+
+func (cursor DistinctCursor) Skip(num int) record.Cursor[record.Record] {
+	cursor.skip += num
+	return cursor
+}
+
+func (cursor DistinctCursor) Limit(num int) record.Cursor[record.Record] {
+	cursor.limit = num
+	return cursor
+}
+
+func (cursor DistinctCursor) Mask(masks ...record.Mask) record.Cursor[record.Record] {
+	cursor.masks = append(cursor.masks, masks...)
+	return cursor
+}
+
+func (cursor DistinctCursor) Sort(sorters ...record.Sorter) record.Cursor[record.Record] {
+	cursor.sorters = sorters
+	return cursor
+}
+
+func (cursor DistinctCursor) Iter() record.Iter[record.Record] {
+	return &DistinctIter{DistinctCursor: cursor}
+}
+
+func (cursor DistinctCursor) Count(ctx context.Context) (int, error) {
+	if len(cursor.keys) == 0 {
+		return 0, nil
+	}
+
+	pipeline := bson.A{
+		bson.M{"$match": makeFilter(cursor.filters)},
+		bson.M{
+			"$group": bson.M{
+				"_id": lo.SliceToMap[string, string, any](
+					cursor.keys,
+					func(key string) (string, any) {
+						return key, fmt.Sprintf("$%s", key)
+					},
+				),
+			},
+		},
+		bson.M{"$group": bson.M{"_id": nil, "totalCount": bson.M{"$sum": 1}}},
+	}
+	result, err := cursor.storage.collection.Aggregate(ctx, pipeline)
+	if err != nil {
+		return 0, fmt.Errorf("failed to aggregate: %s", err)
+	}
+	defer result.Close(ctx)
+
+	var doc struct {
+		TotalCount int `bson:"totalCount"`
+	}
+
+	if !result.Next(ctx) {
+		return 0, nil
+	}
+
+	if err := result.Decode(&doc); err != nil {
+		return 0, fmt.Errorf("failed to decode the document: %s", err)
+	}
+
+	return doc.TotalCount, nil
+}
+
+type DistinctIter struct {
+	DistinctCursor
+	mongoCursor *mongo.Cursor
+}
+
+func (iter *DistinctIter) Close(ctx context.Context) error {
+	if iter.mongoCursor == nil {
+		return nil
+	}
+	return iter.mongoCursor.Close(ctx)
+}
+
+func (iter *DistinctIter) Next(ctx context.Context) (record.Record, error) {
+	if len(iter.keys) == 0 {
+		return nil, record.ErrIterEmpty
+	}
+
+	cursor, err := iter.getCursor(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to make a mongo cursor: %s", err)
+	}
+
+	if !cursor.Next(ctx) {
+		return nil, record.ErrIterEmpty
+	}
+
+	var doc struct {
+		ID bson.M `bson:"_id"`
+	}
+
+	err = cursor.Decode(&doc)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode the received data: %s", err)
+	}
+
+	rec := debsonifyRecord(doc.ID)
+	for _, mask := range iter.masks {
+		if mask.Operation == record.IncludeMask {
+			if _, ok := rec[mask.PropertyPattern]; !ok {
+				rec[mask.PropertyPattern] = nil
```

Comment:
why not delete it?

Reply:
@DenShlk setting included not-existing properties to nil is expected behaviour

Reply:
yup, my bad